cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillController.java

```diff
@@ -484,7 +475,118 @@ private void deletePVCs() {
     return resultFuture;
   }
 
+  /**
+   * Deletes the job controlled by this controller asynchronously.
+   *
+   */
+  private void deleteJob() {
+    apiClient.setDebugging(true);
+    BatchV1Api batchApi = new BatchV1Api(apiClient);
+
+    String name = meta.getName();
+    try {
+      batchApi.deleteNamespacedJobAsync(name, kubeNamespace, null, null, null, null, null,
+                                        new V1DeleteOptions(), new ApiCallbackAdapter<V1Status>() {
+          @Override
+          public void onFailure(ApiException e, int statusCode, Map<String, List<String>> responseHeaders) {
+            // Ignore the failure if the job is already deleted
+            if (statusCode == 404) {
+              LOG.warn("Failed to delete job {}. Job not found.", name, e);
+            }
+          }
+
+          @Override
+          public void onSuccess(V1Status v1Status, int statusCode, Map<String, List<String>> responseHeaders) {
+            deletePods(name);
+          }
+        });
+    } catch (ApiException e) {
+      LOG.warn("Error while cleaning up resources for job {}.", name, e);
```

Comment:
seems like jobs/pods can hang around forever if there is some sort of transient error here?

Do we need some sort of janitor thread that periodically cleans up jobs/pods in case this part isn't able to?

Reply:
Yea, you are right. We noticed same behavior for orphaned dataprocs too. Good to have a cleanup thread.

Reply:
Added a cleanup thread to delete completed jobs.